include/deal.II/base/timer_tree.h

```diff
@@ -0,0 +1,423 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 1998 - 2020 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_timer_tree_h
+#define dealii_timer_tree_h
+
+#include <deal.II/base/conditional_ostream.h>
+#include <deal.II/base/timer.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * A class for wall-time measurements enabling a modular coupling of modules in
+ * a minimally invasive way.
+ */
+class TimerTree
+{
+public:
+  /**
+   * Constructor.
+   */
+  TimerTree()
+    : id("")
+  {}
+
+  /**
+   * Clears the content of this tree. Sub trees inserted
+   * into this tree via pointers to external trees are not touched.
+   */
+  void
+  clear()
+  {
+    this->id = "";
+    data     = nullptr;
+    sub_trees.clear();
+  }
+
+  /**
+   * Inserts a measured @p wall_time into the tree, by
+   * either creating a new entry in the tree if this function is called
+   * the first time with the ID @p ids, or by adding the @p wall_time to an
+   * entry already existing in the tree.
+   */
+  void
+  insert(std::vector<std::string> const ids, double const wall_time)
```

Comment:
```suggestion
  insert(const std::vector<std::string> &ids, const double wall_time)
```

Reply:
This is how I learned C++ ("`const` is a trailing keyword"). Could you please reference me to some coding guidelines stating how deal.II wants to have these things?

Reply:
I believe this not written down explicitly, but it is the de facto convention applied inside deal.II. The coding guidelines state all their examples as a leading keyword, https://dealii.org/developer/doxygen/deal.II/CodingConventions.html , but I think we should state that this is the convention used in that document, to avoid confusion like this.

Reply:
done here https://github.com/dealii/dealii/pull/11939/commits/e7a0fdd8c914cac1eed20326b0a50e94c901bae9

@kronbichler Yes, I suggest that deal.II is more precise about this, e.g. "In deal.II, const is only used as a trailing keyword where necessary. In all other cases that allow to use const either as a trailing or leading keyword, const is used as a leading keyword in deal.II."